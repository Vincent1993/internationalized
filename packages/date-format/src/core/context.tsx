import { createContext, useContext, useMemo } from 'react';
import type { PropsWithChildren } from 'react';
import type { DateFormatContextValue, DateFormatProviderProps } from './types';

const DateFormatContext = createContext<DateFormatContextValue | null>(null);

/**
 * @name DateFormatProvider
 * @description 提供日期时间格式化上下文的 React 组件。它会缓存传入的 `options`
 * 并向所有子组件、Hook 共享，便于在应用内统一 locale、时区等配置。
 *
 * @since 0.0.1
 *
 * @example
 * ```tsx
 * import { DateFormatProvider, DateFormat } from '@internationalized/date-format';
 *
 * export function App() {
 *   return (
 *     <DateFormatProvider options={{ locale: 'zh-CN', timeZone: 'Asia/Shanghai' }}>
 *       <DateFormat value={new Date()} timeStyle="short" dateStyle="medium" />
 *     </DateFormatProvider>
 *   );
 * }
 * ```
 */
export function DateFormatProvider({ children, options }: DateFormatProviderProps) {
  const memoized = useMemo(() => options ?? null, [options]);

  return <DateFormatContext.Provider value={memoized}>{children}</DateFormatContext.Provider>;
}

/**
 * @name useDateFormatContext
 * @description 读取最近的 `DateFormatProvider` 上下文，供自定义组件或 Hook 访问日期配置。
 *
 * @since 0.0.1
 *
 * @returns 当前上下文配置；若不存在 Provider，则返回 `null`
 */
export function useDateFormatContext(): DateFormatContextValue | null {
  return useContext(DateFormatContext);
}

export type { DateFormatContextValue };
export type DateFormatProviderComponentProps = PropsWithChildren<DateFormatProviderProps>;
