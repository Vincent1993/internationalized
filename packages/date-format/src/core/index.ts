export { DateFormatProvider, useDateFormatContext } from './context';
export { createDateFormatter, resolveDateFormatOptions } from './formatter';
export type {
  CreateDateFormatOptions,
  DateFormatContextValue,
  DateFormatter,
  DateFormatProviderProps,
  ResolvedDateFormatOptions,
  UseDateFormatOptions,
  UseDateFormatResult,
  DateInput,
} from './types';
