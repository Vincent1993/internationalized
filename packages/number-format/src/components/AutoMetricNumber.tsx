import { forwardRef, ElementType, memo } from 'react';
import type { ReactNode } from 'react';
import { useAutoFormat } from '../hooks/useAutoFormat';
import type { MetricFormatRule, UseFormatResult } from '../core/types';
import { BaseNumberFormat } from './_internal/BaseNumberFormat';

export type AutoMetricNumberProps<T extends ElementType = 'span'> = Omit<
  React.ComponentPropsWithoutRef<T>,
  'children'
> & {
  value: unknown;
  name: string;
  rules?: MetricFormatRule[];
  fallback?: ReactNode;
  asChild?: boolean;
  as?: T;
  children?: ReactNode | ((result: UseFormatResult) => ReactNode);
};

/**
 * @name AutoMetricNumber
 * @description 一个智能的指标格式化组件，可根据指标名称自动应用预设的格式化规则。
 *   - 核心属性：`value` 数值、`name` 指标名称、`rules` 规则配置
 *   - 渲染属性：`as` 多态元素、`asChild` 插槽模式、`fallback` 错误回退
 *   - 交互属性：`children` 函数式渲染（接收 `UseFormatResult`）
 *   - 支持所有 React 元素属性（继承自 `React.ComponentPropsWithoutRef<T>`）
 *   - 规则可以由 `NumberFormatProvider` 全局提供，也可以由组件局部提供
 *   - 优先级：局部 `rules` > 全局 `rules` > 默认配置
 *
 * @param value - 要格式化的数值
 * @param name - 指标名称，用于匹配格式化规则
 * @param rules - 局部格式化规则数组
 * @param fallback - 错误时的回退内容
 * @param asChild - 是否作为子组件渲染（插槽模式）
 * @param as - 渲染的 HTML 元素类型，默认为 'span'
 * @param children - 函数式子组件，接收 `UseFormatResult` 参数
 * @param ...props - 所有其他 React 元素属性
 *
 * @since 0.0.1
 *
 * @example
 * ```tsx
 * // 1. 在 Provider 中定义全局规则
 * const globalRules = [
 *   { name: 'revenue', options: { style: 'currency', currency: 'USD' } },
 *   { pattern: /rate$/, options: { style: 'percent', maximumFractionDigits: 2 } },
 * ];
 *
 * <NumberFormatProvider options={{ rules: globalRules }}>
 *   <Dashboard />
 * </NumberFormatProvider>
 *
 * // 2. 在 Dashboard 组件中使用
 * function Dashboard() {
 *   return (
 *     <div>
 *       // 匹配到名为 'revenue' 的精确规则
 *       <AutoMetricNumber name="revenue" value={12345.67} />
 *       // => <span>$12,345.67</span>
 *
 *       // 匹配到 `/_rate$/` 的正则规则
 *       <AutoMetricNumber name="conversion_rate" value={0.12345} />
 *       // => <span>12.35%</span>
 *
 *       // 使用局部 `rules` 覆盖全局规则
 *       <AutoMetricNumber
 *         name="revenue"
 *         value={12345.67}
 *         rules={[{ name: 'revenue', options: { currency: 'EUR' } }]}
 *       />
 *       // => <span>€12,345.67</span>
 *
 *       // 使用函数式 children 自定义渲染
 *       <AutoMetricNumber name="growth_rate" value={-0.05}>
 *         {(result) => (
 *           <strong className={result.sign.isNegative ? 'text-red-500' : 'text-green-500'}>
 *             {result.formattedValue}
 *           </strong>
 *         )}
 *       </AutoMetricNumber>
 *       // => <strong class="text-red-500">-5%</strong>
 *     </div>
 *   );
 * }
 * ```
 */
const AutoMetricNumberImpl = forwardRef(
  <T extends ElementType = 'span'>(
    props: AutoMetricNumberProps<T>,
    ref: React.ComponentPropsWithRef<T>['ref'],
  ) => {
    const { name, rules, ...rest } = props;
    const formatter = useAutoFormat({ name, rules });

    return <BaseNumberFormat {...rest} formatter={formatter} ref={ref} />;
  },
);

AutoMetricNumberImpl.displayName = 'AutoMetricNumber';

export const AutoMetricNumber = memo(AutoMetricNumberImpl) as <T extends ElementType = 'span'>(
  props: AutoMetricNumberProps<T> & { ref?: React.ComponentPropsWithRef<T>['ref'] },
) => JSX.Element;
