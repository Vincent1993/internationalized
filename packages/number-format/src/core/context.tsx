import { createContext, useContext, useMemo } from 'react';
import type { NumberFormatContextValue, NumberFormatProviderProps } from './types';
import { mergeRules } from './utils/rules';

const NumberFormatContext = createContext<NumberFormatContextValue | null>(null);

/**
 * @name NumberFormatProvider
 * @description 为应用提供全局的数字格式化配置上下文，支持所有 `Intl.NumberFormat`
 * 选项与指标规则合并。
 *
 * @since 0.0.1
 *
 * @example
 * ```tsx
 * import { NumberFormatProvider, NumberFormat, AutoMetricNumber } from '@your-package/number-format';
 *
 * // 推荐用法：使用 options 对象
 * function App() {
 *   return (
 *     <NumberFormatProvider
 *       options={{
 *         locale: "zh-CN",
 *         useGrouping: true,
 *         minimumFractionDigits: 2
 *       }}
 *     >
 *       <Dashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 带指标规则的配置
 * function AppWithRules() {
 *   const options = {
 *     locale: "zh-CN",
 *     useGrouping: true,
 *     rules: [
 *       // 精确匹配规则
 *       { name: 'revenue', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'profit', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'cost', options: { style: 'currency', currency: 'CNY' } },
 *
 *       // 正则匹配规则
 *       { pattern: '.*_rate$', options: { style: 'percent', maximumFractionDigits: 2 } },
 *       { pattern: '.*_ratio$', options: { style: 'percent', maximumFractionDigits: 1 } },
 *       { pattern: '.*_count$', options: { style: 'decimal', useGrouping: true } },
 *       { pattern: '.*_permille$', options: { style: 'per-mille' } },
 *       { pattern: '.*_permyriad$', options: { style: 'per-myriad' } },
 *       { pattern: '.*_pp$', options: { style: 'percentage-point' } },
 *       { pattern: '.*_cn_upper$', options: { style: 'cn-upper' } },
 *     ]
 *   };
 *
 *   return (
 *     <NumberFormatProvider options={options}>
 *       <MetricDashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 嵌套 Provider（局部覆盖全局配置）
 * function NestedProviders() {
 *   return (
 *     <NumberFormatProvider options={{ locale: "zh-CN", useGrouping: true }}>
 *       <div>
 *         <h1>Global Config</h1>
 *         <NumberFormat value={1234.56} />
 *
 *         <NumberFormatProvider options={{ locale: "en-US", style: "currency", currency: "USD" }}>
 *           <div>
 *             <h2>Local Override</h2>
 *             <NumberFormat value={1234.56} />
 *           </div>
 *         </NumberFormatProvider>
 *       </div>
 *     </NumberFormatProvider>
 *   );
 * }
 * ```
 */
export const NumberFormatProvider = ({ children, options }: NumberFormatProviderProps) => {
  const parentValue = useNumberFormatContext();

  const value = useMemo<NumberFormatContextValue | null>(() => {
    if (!options) {
      return parentValue;
    }

    if (!parentValue) {
      return options;
    }

    const mergedValue: NumberFormatContextValue = {
      ...parentValue,
      ...options,
      rules: mergeRules(parentValue.rules || [], options.rules || []),
    };

    if (mergedValue.rules?.length === 0) {
      delete mergedValue.rules;
    }

    return mergedValue;
  }, [parentValue, options]);

  return <NumberFormatContext.Provider value={value}>{children}</NumberFormatContext.Provider>;
};

/**
 * @name useNumberFormatContext
 * @description React Hook，用于获取当前的数字格式化上下文配置，主要供内部 Hook
 * 与组件使用，也可用于自定义封装。
 *
 * @since 0.0.1
 *
 * @returns 当前上下文配置；若不存在 Provider 则返回 `null`
 */
export const useNumberFormatContext = (): NumberFormatContextValue | null => {
  return useContext(NumberFormatContext);
};
