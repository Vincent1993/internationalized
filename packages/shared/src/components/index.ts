export { Slot } from './Slot';
export { createBaseFormatComponent } from './createBaseFormatComponent';
export type { BaseFormatComponentProps, FormatInvoker } from './createBaseFormatComponent';
