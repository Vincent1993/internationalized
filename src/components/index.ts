// Export components
/* v8 ignore next */
export { NumberFormat } from './NumberFormat';
export { AutoMetricNumber } from './AutoMetricNumber';

// Export component types
export type { AutoMetricNumberProps } from './AutoMetricNumber';
