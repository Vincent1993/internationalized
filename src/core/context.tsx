import { createContext, useContext, useMemo } from 'react';
import type { NumberFormatContextValue, NumberFormatProviderProps } from './types';
import { mergeRules } from './utils/rules';

const NumberFormatContext = createContext<NumberFormatContextValue | null>(null);

/**
 * 全局数字格式化 Provider
 *
 * @description 为应用提供全局的数字格式化配置上下文，支持所有 Intl.NumberFormat 选项和指标规则
 *
 * @example
 * ```tsx
 * import { NumberFormatProvider, NumberFormat, AutoMetricNumber } from '@your-package/number-format';
 *
 * // 推荐用法：使用 options 对象
 * function App() {
 *   return (
 *     <NumberFormatProvider
 *       options={{
 *         locale: "zh-CN",
 *         useGrouping: true,
 *         minimumFractionDigits: 2
 *       }}
 *     >
 *       <Dashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 带指标规则的配置
 * function AppWithRules() {
 *   const options = {
 *     locale: "zh-CN",
 *     useGrouping: true,
 *     rules: [
 *       // 精确匹配规则
 *       { name: 'revenue', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'profit', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'cost', options: { style: 'currency', currency: 'CNY' } },
 *
 *       // 正则匹配规则
 *       { pattern: '.*_rate$', options: { style: 'percent', maximumFractionDigits: 2 } },
 *       { pattern: '.*_ratio$', options: { style: 'percent', maximumFractionDigits: 1 } },
 *       { pattern: '.*_count$', options: { style: 'decimal', useGrouping: true } },
 *       { pattern: '.*_permille$', options: { style: 'per-mille' } },
 *       { pattern: '.*_permyriad$', options: { style: 'per-myriad' } },
 *       { pattern: '.*_pp$', options: { style: 'percentage-point' } },
 *       { pattern: '.*_cn_upper$', options: { style: 'cn-upper' } },
 *     ]
 *   };
 *
 *   return (
 *     <NumberFormatProvider options={options}>
 *       <MetricDashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 嵌套 Provider（局部覆盖全局配置）
 * function NestedProviders() {
 *   return (
 *     <NumberFormatProvider options={{ locale: "zh-CN", useGrouping: true }}>
 *       <div>
 *         <h1>Global Config</h1>
 *         <NumberFormat value={1234.56} />
 *
 *         <NumberFormatProvider options={{ locale: "en-US", style: "currency", currency: "USD" }}>
 *           <div>
 *             <h2>Local Override</h2>
 *             <NumberFormat value={1234.56} />
 *           </div>
 *         </NumberFormatProvider>
 *       </div>
 *     </NumberFormatProvider>
 *   );
 * }
 * ```
 */
export const NumberFormatProvider = ({ children, options }: NumberFormatProviderProps) => {
  const parentValue = useNumberFormatContext();

  const value = useMemo<NumberFormatContextValue | null>(() => {
    if (!options) {
      return parentValue;
    }

    if (!parentValue) {
      return options;
    }

    const mergedValue: NumberFormatContextValue = {
      ...parentValue,
      ...options,
      rules: mergeRules(parentValue.rules || [], options.rules || []),
    };

    if (mergedValue.rules?.length === 0) {
      delete mergedValue.rules;
    }

    return mergedValue;
  }, [parentValue, options]);

  return <NumberFormatContext.Provider value={value}>{children}</NumberFormatContext.Provider>;
};

/**
 * 获取数字格式化上下文
 * @internal
 * @description React Hook，用于获取当前的数字格式化上下文配置
 * 主要供内部 hooks 和组件使用，也可用于自定义组件
 *
 * @returns 当前上下文的配置值，如果没有 Provider 则返回空对象
 *
 */
export const useNumberFormatContext = (): NumberFormatContextValue | null => {
  return useContext(NumberFormatContext);
};
