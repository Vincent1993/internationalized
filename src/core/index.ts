/**
 * @file 核心功能公共 API
 */

// ==================== 核心 API ====================
export { NumberFormatProvider } from './context';
export { createNumberFormat, resolveFormatOptions } from './formatter';

// ==================== 核心类型 ====================
export type {
  MetricFormatRule,
  NumberFormatContextValue,
  UseFormatOptions,
  UseFormatResult,
  CreateNumberFormatOptions,
  NumberFormatter,
  ResolvedFormatOptions,
} from './types';
