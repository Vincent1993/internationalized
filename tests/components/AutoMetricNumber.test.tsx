import { render, screen } from '@testing-library/react';
import React from 'react';
import { AutoMetricNumber } from '../../src/components/AutoMetricNumber';
import { NumberFormatProvider } from '../../src/core/context';
import type { MetricFormatRule } from '../../src/core/types';

describe('components/AutoMetricNumber', () => {
  const mockRules: MetricFormatRule[] = [
    {
      pattern: /percentage|percent|rate/i,
      options: { style: 'percent', maximumFractionDigits: 2 },
    },
    {
      pattern: /price|cost|amount|money/i,
      options: { style: 'currency', currency: 'CNY' },
    },
    {
      pattern: /count|number|total/i,
      options: { style: 'decimal', maximumFractionDigits: 0 },
    },
  ];

  it('应该根据指标名称自动应用格式化规则', () => {
    render(<AutoMetricNumber name='conversion_percentage' value={0.1234} rules={mockRules} />);

    expect(screen.getByText('12.34%')).toBeInTheDocument();
  });

  it('应该支持货币格式化', () => {
    render(
      <AutoMetricNumber
        name='product_price'
        value={199.99}
        rules={mockRules}
        data-testid='currency-format'
      />,
    );

    const element = screen.getByTestId('currency-format');
    expect(element.textContent).toMatchSnapshot();
  });

  it('应该支持整数计数格式化', () => {
    render(<AutoMetricNumber name='user_count' value={1234.56} rules={mockRules} />);

    expect(screen.getByText('1,235')).toBeInTheDocument(); // 四舍五入到整数
  });

  it('应该在无匹配规则时使用默认配置', () => {
    render(<AutoMetricNumber name='unknown_metric' value={123.456} rules={mockRules} />);

    expect(screen.getByText('123.456')).toBeInTheDocument();
  });

  it('应该支持asChild模式', () => {
    render(
      <AutoMetricNumber name='test_metric' value={123.45} rules={mockRules} asChild>
        {(result) => <span data-testid='auto-child'>{result.formattedValue}</span>}
      </AutoMetricNumber>,
    );

    const element = screen.getByTestId('auto-child');
    expect(element).toHaveTextContent('123.45');
    expect(element.tagName).toBe('SPAN');
  });

  describe('规则匹配测试', () => {
    it('应该匹配复杂的正则表达式模式', () => {
      const complexRules: MetricFormatRule[] = [
        {
          pattern: /^(room|outdoor|indoor)_temp(erature)?$/i,
          options: { style: 'unit', unit: 'celsius' },
        },
        {
          pattern: /(speed|velocity)_(mph|kmh|ms)$/i,
          options: { style: 'unit', unit: 'meter-per-second' },
        },
      ];

      render(
        <AutoMetricNumber
          name='room_temp'
          value={23.5}
          rules={complexRules}
          data-testid='temp-format'
        />,
      );

      const element = screen.getByTestId('temp-format');
      expect(element.textContent).toMatchSnapshot();
    });

    it('应该使用最后一个匹配的规则', () => {
      const overlappingRules: MetricFormatRule[] = [
        {
          pattern: /rate/i,
          options: { style: 'percent', maximumFractionDigits: 1 },
        },
        {
          pattern: /rate/i,
          options: { style: 'percent', maximumFractionDigits: 3 },
        },
      ];

      render(
        <AutoMetricNumber
          name='conversion_rate'
          value={0.12345}
          rules={overlappingRules}
          data-testid='first-match'
        />,
      );

      // 应该使用第一个规则（1位小数）
      expect(screen.getByTestId('first-match')).toHaveTextContent('12.345%');
    });

    it('应该区分大小写敏感性', () => {
      const caseSensitiveRules: MetricFormatRule[] = [
        {
          pattern: /TEST/,
          options: { style: 'decimal', maximumFractionDigits: 0 },
        },
      ];

      render(
        <AutoMetricNumber
          name='test_metric'
          value={123.45}
          rules={caseSensitiveRules}
          data-testid='case-test'
        />,
      );

      // 应该使用默认配置（2位小数）而不是规则（0位小数）
      expect(screen.getByTestId('case-test')).toHaveTextContent('123.45');
    });
  });

  describe('与Context集成测试', () => {
    const contextRules: MetricFormatRule[] = [
      {
        pattern: /global_rate/i,
        options: { style: 'percent', maximumFractionDigits: 1 },
      },
    ];

    const wrapper = ({ children }: { children: React.ReactNode }) => (
      <NumberFormatProvider
        options={{
          locale: 'en-US',
          rules: contextRules,
          useGrouping: false,
        }}
      >
        {children}
      </NumberFormatProvider>
    );

    it('应该合并Context和局部规则', () => {
      const localRules: MetricFormatRule[] = [
        {
          pattern: /local_price/i,
          options: { style: 'currency', currency: 'USD' },
        },
      ];

      render(
        <AutoMetricNumber
          name='global_rate'
          value={0.123}
          rules={localRules}
          data-testid='context-rule'
        />,
        { wrapper },
      );

      // 应该匹配Context中的规则
      expect(screen.getByTestId('context-rule')).toHaveTextContent('12.3%');
    });

    it('应该局部规则优先于Context规则', () => {
      const localRules: MetricFormatRule[] = [
        {
          pattern: /global_rate/i,
          options: { style: 'percent', maximumFractionDigits: 3 }, // 覆盖Context的1位小数
        },
      ];

      render(
        <AutoMetricNumber
          name='global_rate'
          value={0.12345}
          rules={localRules}
          data-testid='local-override'
        />,
        { wrapper },
      );

      // 应该使用局部规则（3位小数）
      expect(screen.getByTestId('local-override')).toHaveTextContent('12.345%');
    });
  });

  describe('props传递测试', () => {
    it('应该将额外的props传递给底层组件', () => {
      render(
        <AutoMetricNumber
          name='test_metric'
          value={123}
          rules={mockRules}
          className='custom-class'
          id='test-id'
          title='Test title'
          data-testid='props-forward'
        />,
      );

      const element = screen.getByTestId('props-forward');
      expect(element).toHaveClass('custom-class');
      expect(element).toHaveAttribute('id', 'test-id');
      expect(element).toHaveAttribute('title', 'Test title');
    });

    it('应该支持ref转发', () => {
      const ref = React.createRef<HTMLSpanElement>();

      render(<AutoMetricNumber name='test_metric' value={123} rules={mockRules} ref={ref} />);

      expect(ref.current).toBeInstanceOf(HTMLSpanElement);
      expect(ref.current?.textContent).toBe('123');
    });
  });

  describe('快照测试', () => {
    it('应该为不同指标类型生成正确的快照', () => {
      const testCases = [
        { name: 'conversion_percentage', value: 0.1234 },
        { name: 'product_price', value: 299.99 },
        { name: 'user_count', value: 1234.56 },
        { name: 'unknown_metric', value: 42.195 },
        { name: 'click_through_rate', value: 0.0567 },
        { name: 'average_session_duration', value: 234.67 },
      ];

      testCases.forEach(({ name, value }) => {
        const { container } = render(
          <AutoMetricNumber name={name} value={value} rules={mockRules} />,
        );

        expect(container.firstChild).toMatchSnapshot(`auto-metric-${name}`);
      });
    });

    it('应该为不同的规则配置生成快照', () => {
      const ruleVariations = [
        [], // 空规则
        mockRules, // 基础规则
        [
          ...mockRules,
          {
            pattern: /temp|temperature/i,
            options: { style: 'unit', unit: 'celsius' },
          },
        ], // 扩展规则
      ];

      ruleVariations.forEach((rules, index) => {
        const { container } = render(
          <AutoMetricNumber name='test_metric' value={25.5} rules={rules} />,
        );

        expect(container.firstChild).toMatchSnapshot(`rules-variation-${index}`);
      });
    });
  });
});
