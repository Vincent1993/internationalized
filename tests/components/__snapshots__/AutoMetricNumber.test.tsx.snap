// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/AutoMetricNumber > 应该支持货币格式化 1`] = `"¥199.99"`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同指标类型生成正确的快照 > auto-metric-average_session_duration 1`] = `
<span>
  234.67
</span>
`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同指标类型生成正确的快照 > auto-metric-click_through_rate 1`] = `
<span>
  5.67%
</span>
`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同指标类型生成正确的快照 > auto-metric-conversion_percentage 1`] = `
<span>
  12.34%
</span>
`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同指标类型生成正确的快照 > auto-metric-product_price 1`] = `
<span>
  ¥299.99
</span>
`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同指标类型生成正确的快照 > auto-metric-unknown_metric 1`] = `
<span>
  42.195
</span>
`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同指标类型生成正确的快照 > auto-metric-user_count 1`] = `
<span>
  1,235
</span>
`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同的规则配置生成快照 > rules-variation-0 1`] = `
<span>
  25.5
</span>
`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同的规则配置生成快照 > rules-variation-1 1`] = `
<span>
  25.5
</span>
`;

exports[`components/AutoMetricNumber > 快照测试 > 应该为不同的规则配置生成快照 > rules-variation-2 1`] = `
<span>
  25.5
</span>
`;

exports[`components/AutoMetricNumber > 规则匹配测试 > 应该匹配复杂的正则表达式模式 1`] = `"23.5°C"`;
